HLRC: Add ILM Status to HLRC

jakelandis opened the pull request:
Adds support for the Index Lifecycle Management Status to the Java
High-Level Rest Client.

Relates to #33100

--------
Notes to reviewer(s):
* I don't think there should be anything controversial here, and probably don't need 5 reviewers ... just wasn't sure who to tag. 
* the naming conventions were chosen to match the start/stopILM.
* Using a TimedRequest instead of a StatusILMRequest... not sure the preference here, but it kinda matches the `AcknowledgedResponse` in terms of not creating typed request/response when not needed.

elasticmachine commented:
Pinging @elastic/es-core-infra

gwbrown commented on client/rest-high-level/src/main/java/org/elasticsearch/client/IndexLifecycleClient.java:
```diff
@@ -139,6 +140,34 @@ public AcknowledgedResponse stopILM(StopILMRequest request, RequestOptions optio
+     * @param request the request with user defined timeouts.
+     * @param options the request options (e.g. headers), use {@link RequestOptions#DEFAULT} if nothing needs to be customized
+     */
+    public StatusILMResponse StatusILM(TimedRequest request, RequestOptions options) throws IOException {
```
Method names should start with a lowercase letter, so this should be `statusILM`.

Or maybe `getILMStatus`? `statusILM` doesn't make much sense seeing it for the first time - there's no verb there.

gwbrown commented on client/rest-high-level/src/main/java/org/elasticsearch/client/IndexLifecycleClient.java:
```diff
@@ -139,6 +140,34 @@ public AcknowledgedResponse stopILM(StopILMRequest request, RequestOptions optio
+     * @param options  the request options (e.g. headers), use {@link RequestOptions#DEFAULT} if nothing needs to be customized
+     * @param listener the listener to be notified upon request completion
+     */
+    public void StatusILMAsync(TimedRequest request, RequestOptions options,
```
As above re: method names.

dakrone commented on client/rest-high-level/src/main/java/org/elasticsearch/client/IndexLifecycleClient.java:
```diff
@@ -139,6 +140,34 @@ public AcknowledgedResponse stopILM(StopILMRequest request, RequestOptions optio
+     * @param request the request with user defined timeouts.
+     * @param options the request options (e.g. headers), use {@link RequestOptions#DEFAULT} if nothing needs to be customized
+     */
+    public StatusILMResponse StatusILM(TimedRequest request, RequestOptions options) throws IOException {
```
I think these should be `LifecycleManagementStatusResponse`, rather than using the ILM in method/class names

Also, the method name is `StatusILM` and we don't capitalize any of our method names

jakelandis replied:
> Also, the method name is StatusILM and we don't capitalize any of our method names

oops didn't meant to capitalize the method name. (kinda surprised that check style didn't catch that). will fix. 

> I think these should be LifecycleManagementStatusResponse, rather than using the ILM in method/class names

I tend to agree, but was matching the start/stopILM here:  https://github.com/elastic/elasticsearch/tree/index-lifecycle/x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/indexlifecycle 

Should I rename those too ?

dakrone commented on client/rest-high-level/src/main/java/org/elasticsearch/client/IndexLifecycleClient.java:
```diff
@@ -139,6 +140,34 @@ public AcknowledgedResponse stopILM(StopILMRequest request, RequestOptions optio
+     * @param options  the request options (e.g. headers), use {@link RequestOptions#DEFAULT} if nothing needs to be customized
+     * @param listener the listener to be notified upon request completion
+     */
+    public void StatusILMAsync(TimedRequest request, RequestOptions options,
```
Same here about naming and capitalization

jakelandis replied:
oops ... IDE find and replace gone wild. will fix.

dakrone commented on client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/StatusILMResponse.java:
```diff
@@ -0,0 +1,99 @@
+        }
+    }
+
+    // generated
```
I don't think we need these generated comments

jakelandis replied:
I can remove them.  FWIW, I generally put these in here as a way to signal nothing special here, so no tests, and if changes are made need to re-generate.

gwbrown commented on client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/StatusILMResponse.java:
```diff
@@ -0,0 +1,99 @@
+    /**
+     * Enum representing the different modes that Index Lifecycle Service can operate in.
+     */
+    public enum OperationMode {
```
#33250 moves the existing `OperationMode` class into `o.e.client.indexlifecycle` - you may want to wait until that gets merged (should be later today) and use that one instead of having this as an inner class.

jakelandis replied:
thanks! ... will wait till that gets merged and update accordingly.

gwbrown commented on client/rest-high-level/src/test/java/org/elasticsearch/client/IndexLifecycleIT.java:
```diff
@@ -169,35 +169,30 @@ public void testStartStopILM() throws Exception {
-        Response statusResponse = client().performRequest(statusReq);
-        String statusResponseString = EntityUtils.toString(statusResponse.getEntity());
-        assertEquals("{\"operation_mode\":\"RUNNING\"}", statusResponseString);
+        TimedRequest statusRequest = new TimedRequest();
```
While I get the logic behind why you did it this way, I think generic classes make more sense for responses than requests.  As a user of the HLRC, this feels really strange - even if `GetILMStatusRequest` would be an extra class that doesn't really add anything, I think having separate classes for each request type is the way to go.  I could be wrong though, I'd like to hear others' input on this.

jakelandis replied:
I don't have a strong opinion beyond consistency in the API.  

Perhaps we should allow a no-arg method in the client that creates the request for them. However, that also creates inconsistency (unless you create a bunch of new methods).

> I'd like to hear others' input on this.

..same

gwbrown replied:
@hub-cap Could you weigh in on this?

gwbrown commented on client/rest-high-level/src/test/java/org/elasticsearch/client/indexlifecycle/StatusILMResponseTests.java:
```diff
@@ -0,0 +1,66 @@
+
+public class StatusILMResponseTests extends ESTestCase {
+
+    public void testClientServerStatuses() {
```
Using the shared class would let you get rid of this test, too.

jakelandis replied:
will wait for #33250 to use shared class.

jakelandis commented:
Changes pushed:
* Fixed incorrect method uppercase 
* Renamed statsILM -> lifecycleManagementStatus
* removed "generated" comments

TODO (this PR):
* Integrate with #33250 once merged
* Decision on named Request classes vs. generic TimedRequest class for empty requests [1]

TODO (not in this PR):
* rename stopILM and startILM ? [2]
* enforce basic method naming via checkstyle

@hub-cap [1]
Is there a preferred design for how to model empty requests (no parameters) for the API ? 
a) Use a generic TimedRequest (like this PR currently)
b) create an empty typed Request (e.g. `public class MyThingRequest extends TimedRequest{ //empty }` )  (as @gwbrown advocates)
c) introduce no arg methods in the *Client's that under the covers uses a TimedRequest (e.g. `MyThingResponse getMyThing(){ //doit }`)
d) other ?


@colings86 [2]
Any objections if I move the start/stopILM HLRC API's to the newer format and change the name to match the @dakrone's naming suggestion ? (different PR from this one)

gwbrown commented:
@jakelandis #33250 has been merged into `index-lifecycle` and backported to `index-lifecycle-6.x`, so you should be good to go for switching over to the existing `OperationMode` class.

jakelandis commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java:
```diff
@@ -1241,6 +1241,18 @@ static Request stopILM(StopILMRequest stopILMRequest) {
         return request;
     }
 
+    static Request statusILM(TimedRequest ilmStatusRequest){
```
thanks missed this one. done.

gwbrown commented:
LGTM, pending @hub-cap's input on specialized vs generic requests.

hub-cap commented:
If we create a request and we ever want to add things to it, then it will not require a breaking contract change. It also gives us a chance to rename TimedRequest if its scope changes one day. It might even be worth making it abstract to ensure we dont use it this way. what do yall think?

jakelandis commented:
@hub-cap - good point. I will add the specific the request to this PR and change TimedRequest to abstract.

jakelandis commented:
@dakrone @gwbrown - mind taking another look all suggestions are implemented.

dakrone commented on client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/LifecycleManagementStatusResponse.java:
```diff
@@ -0,0 +1,67 @@
+        "operation_mode", a -> new LifecycleManagementStatusResponse((String) a[0]));
+
+    static {
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), new ParseField("operation_mode"));
```
Can you put the `ParseField` into a class private var and then use it in the parser (so we don't accidentally typo/change it in the future)

jakelandis replied:
moved "operation_mode" to a static final string (not exactly the ask, but i believe same outcome)

dakrone commented on client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/LifecycleManagementStatusResponse.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    //package private for testing
+    LifecycleManagementStatusResponse(String operationMode) {
```
I think this should take an `OperationMode` to push the conversion into the caller, rather than doing it in the constructor (it seems cleaner to me that way).

This is just my personal preference though, so up to you if you want to change it.

jakelandis replied:
leaving as-is

dakrone commented on client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/OperationMode.java:
```diff
@@ -56,4 +59,10 @@ public boolean isValidChange(OperationMode nextMode) {
+    static OperationMode fromString(String string) {
+        return EnumSet.allOf(OperationMode.class).stream()
+            .filter(e -> string.equalsIgnoreCase(e.name())).findFirst()
+            .orElseThrow(() -> new IllegalArgumentException(String.format(Locale.ENGLISH, "%s is not a valid operation_mode", string)));
```
We use `Locale.ROOT` for as many things as we can, rather than the English locale here

jakelandis replied:
done

jakelandis commented:
Thanks for the reviews!  will merge once CI turns green.

jakelandis commented:
jenkins test this please
